### PR TITLE
Add CanIUse basic search support as a chat plugin.

### PR DESCRIPTION
**Pull request by Nessworthy**
As a basic level plugin, can I use adds two command aliases: `!!ciu` and `!!caniuse`.
Usage:
* `!!caniuse` - Prints out a simple link to http://caniuse.com.
* `!!caniuse [search term]` - Prints out a caniuse link for that search term.

Future areas of development include scraping the page to display more accurate information (e.g. search term not found, or to display a summary of the top search result). Unfortunately, I can't seem to find an API for the resource at present.

Requires testing in playground.

Had to rebase because of some author issues. If you value a clean repo I don't mind submitting this under a new fork.

### CODE REVIEW

File: src/Chat/Plugin/CanIUse.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * Get a list of specific commands handled by this plugin
+     *
+     * @return string[]
+     */
+    public function getHandledCommands(): array
+    {
+        return ['caniuse', 'ciu'];
```
Comment (PeeHaa) [inline src/Chat/Plugin/CanIUse.php]: Perhaps we should just use a single command here and alias it at runtime instead? cc @DaveRandom

Comment (kelunik) [inline src/Chat/Plugin/CanIUse.php]: Yes, should be a single command and later mapped.

**Comment by PeeHaa** [comment]
You need to implement to `\Room11\Jeeves\Chat\Plugin` methods :-)

**Comment by DaveRandom** [comment]
The raw data is available at https://github.com/fyrd/caniuse... we'd have to index it somehow or work out a way to search by scraping

**Comment by Nessworthy** [comment]
@PeeHaa Not sure what to implement for `handleMessage`. `handlesAllMessages` will return false in my next commit, but I'm not sure what `\Generator` I should be returning for the former method

**Comment by DaveRandom** [comment]
```php
class CanIUse implements Plugin
{
    // there are a few other things we can do with plugins that we don't need to worry about here
    // so this trait just hides those interface implementation methods
    use Plugin\Traits\CommandOnly;

    const DOMAIN = 'http://caniuse.com';

    private $chatClient;

    public function __construct(ChatClient $chatClient)
    {
        $this->chatClient = $chatClient;
    }

    /**
     * Generate a link for Can I Use, based on the search term.
     *
     * @param  string $searchFor The search term, if any.
     * @return string            A markdown formatted link for Can I Use.
     */
    private function generateLink(string $searchFor): string
    {
        $searchFor = trim($searchFor);
        $title = 'Can I Use - Support tables for HTML5, CSS3, etc';
        $url   = self::DOMAIN;
        if (strlen($searchFor) > 0) {
            $title = sprintf('Can I Use - `%s`', ucwords($searchFor));
            $url   = $this->generateSearchUri($searchFor);
        }
        return sprintf('[%s](%s)', $title, $url);
    }
    /**
     * Create a search URI for a given search term.
     *
     * @param  string $searchFor The search string.
     * @return string            The full search URI.
     */
    private function generateSearchUri(string $searchFor)
    {
        return self::DOMAIN . '/' . urlencode($searchFor);
    }

    /**
     * Entry point for the Link endpoint
     *
     * If an entry point function returns a generator it will be treated as a co-routine, in this case
     * you can just return the promise directly. Any other return value will be ignore, so if your plugin
     * doesn't need to run anything async just don't bother returning anything.
     *
     * @param Command $command
     * @return Promise
     */
    public function getLink(Command $command): Promise
    {
        return $this->chatClient->postReply($command, $this->generateLink(implode(' ', $command->getParameters())));
    }

    public function getName(): string
    {
        return 'CanIUse'; // a unique (case insensitive) name for the plugin
    }

    public function getDescription(): string
    {
        return 'A human readable description of what the plugin does';
    }

    public function getHelpText(array $args): string
    {
        // TODO: Implement getHelpText() method.
        // this currently does nothing so you can ignore it, I just leave the todo comment there
    }

    /**
     * @return PluginCommandEndpoint[]
     */
    public function getCommandEndpoints(): array
    {
        return [
            /* a plugin now exposes "endpoints" rather than just registering specific commands to handle
               this is how the command remapping etc works */
            new PluginCommandEndpoint(
                'Link', // the unique, descriptive name of this endpoint
                [$this, 'getLink'], // a callable that will be invoked with a command
                'caniuse' // the *default* command for this endpoint (but can be remapped or disabled)
                // you can also pass a 4th argument which is a description of the endpoint, which defaults
                // to the plugin description. When there's only one endpoint there's usually no point.
            ),
        ];
    }
}
```

**Pull request merged**